src/sourcegraph.d.ts

```diff
@@ -338,42 +337,251 @@ declare module 'sourcegraph' {
+         * Updates the configuration value for the given key. The updated configuration value is persisted by the
+         * client.
+         *
+         * @template C The configuration schema.
+         * @template K Valid key on the configuration object.
+         * @param key The name of the configuration property to update.
+         * @param value The new value, or undefined to remove it.
+         * @return A promise that resolves when the client acknowledges the update.
+         */
+        update<K extends keyof C>(key: K, value: C[K] | undefined): Promise<void>
+
+        /**
+         * The configuration value as a plain object.
+         */
+        readonly value: Readonly<C>
+    }
+
+    /**
+     * The configuration settings.
      *
-     * ```ts
-     * let a: HoverProvider = {
-     * 	provideHover(doc, pos, token): ProviderResult<Hover> {
-     * 		return new Hover('Hello World');
-     * 	}
-     * }
+     * It may be merged from the following sources of settings, in order:
      *
-     * let b: HoverProvider = {
-     * 	provideHover(doc, pos, token): ProviderResult<Hover> {
-     * 		return new Promise(resolve => {
-     * 			resolve(new Hover('Hello World'));
-     * 	 	});
-     * 	}
-     * }
+     * Default settings
+     * Global settings
+     * Organization settings (for all organizations the user is a member of)
+     * User settings
+     * Client settings
+     * Repository settings
+     * Directory settings
      *
-     * let c: HoverProvider = {
-     * 	provideHover(doc, pos, token): ProviderResult<Hover> {
-     * 		return; // undefined
-     * 	}
-     * }
-     * ```
+     * @todo Add a way to get/update configuration for a specific scope or subject.
+     * @todo Support applying defaults to the configuration values.
      */
-    export type ProviderResult<T> = T | undefined | null | Promise<T | undefined | null>
+    export namespace configuration {
+        /**
+         * Returns the full configuration object.
+         *
+         * @todo This function throws an error if it is called synchronously in the extension's `activate`
+         *       function. This will be fixed before beta. See the test "Configuration (integration) / is usable in
+         *       synchronous activation functions".
+         *
+         * @template C The configuration schema.
+         * @return The full configuration object.
+         */
+        export function get<C = { [key: string]: any }>(): Configuration<C>
+
+        /**
+         * Subscribe to changes to the configuration. The {@link next} callback is called when any configuration
+         * value changes (and synchronously immediately). Call {@link get} in the callback to obtain the new
+         * configuration values.
+         *
+         * @template C The configuration schema.
+         * @return An unsubscribable to stop calling the callback for configuration changes.
+         */
+        export function subscribe(next: () => void): Unsubscribable
```

Comment:
Does this allow `from(sourcegraph.configuration).pipe()` if it doesn't have onerror/oncomplete?

Reply:
No. I would love your advice on how to make it support that, without complicating this API. (If you're not using observables, there is very little value in the onerror and oncomplete callbacks, so I would like to keep them out of this type signature.)